rustls/src/server/mod.rs

```diff
@@ -146,6 +146,311 @@ impl<'a> ClientHello<'a> {
     }
 }
 
+/// Building a ServerConfig in a linker-friendly way.
+///
+/// Linker-friendly: meaning unused ciphersuites, protocol
+/// versions, key exchange mechanisms, etc. can be discarded
+/// by the linker as they'll be unreferenced.
+///
+/// Example:
+///
+/// ```ignore
+/// ServerConfigBuilder::new()
+///     .with_default_ciphersuites()
+///     .with_default_kx_groups()
+///     .without_client_auth()
+///     .with_single_cert(certs, private_key)
+///     .build();
+/// ```
+///
+/// This may be shortened to:
+/// 
+/// ```ignore
+/// ServerConfigBuilder::defaults()
+///     .without_client_auth()
+///     .with_single_cert(certs, private_key)
+///     .build();
+/// ```
+///
+/// The types used here fit together like this:
+///
+/// 1. Get a ServerConfigBuilder with `ServerConfigBuilder::new()`.
+///    You must make a decision on which ciphersuites to use, typically
+///    by calling `with_default_ciphersuites()`.
+/// 2. You now have a ServerConfigBuilderWithSuites.  You must make a decision
+///    on key exchange groups: typically by calling `with_default_kx_groups()`.
+/// 2. You now have a ServerConfigBuilderWithSuitesAndKx.  You must make
+///    a decision on how and whether to use client authentication.  Perhaps
+///    you call `without_client_auth()`.
+/// 3. You now have a ServerConfigBuilderComplete.  This object has a number
+///    of defaults you can change, or you can pass it to `ServerConfig::new`.
+pub struct ServerConfigBuilder {}
+
+impl ServerConfigBuilder {
+    /// Start building a ServerConfig.
+    pub fn new() -> ServerConfigBuilder {
+        ServerConfigBuilder {}
+    }
+
+    /// Start building a ServerConfig, and accept defaults for underlying
+    /// cryptography.
+    ///
+    /// These are safe defaults, useful for 99% of applications.
+    ///
+    /// With the returned object, you must still make decisions on
+    /// client authentication and provide server credentials -- rustls
+    /// can't provide useful defaults for these.
+    pub fn defaults() -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilder::new()
+            .with_default_ciphersuites()
+            .with_default_kx_groups()
+    }
+
+    /// Choose a specific set of ciphersuites.
+    pub fn with_ciphersuites(
+        self,
+        ciphersuites: &[&'static SupportedCipherSuite],
+    ) -> ServerConfigBuilderWithSuites {
+        ServerConfigBuilderWithSuites {
+            ciphersuites: ciphersuites.to_vec(),
+        }
+    }
+
+    /// Choose the default set of ciphersuites.
+    ///
+    /// Note that this default provides only high-quality suites: there is no need
+    /// to filter out low-, export- or NULL-strength ciphersuites: rustls does not
+    /// implement these.
+    pub fn with_default_ciphersuites(self) -> ServerConfigBuilderWithSuites {
```

Comment:
I think it would be simpler to just have callers call `with_ciphersuites(DEFAULT_CIPHERSUITES)` directly, and recommend they pass `DEFAULT_CIPHERSUITES` unless they know what they're doing. Similarly for similar methods.

Reply:
It's a trade-off between downstream users having to reference/import the extra symbol vs maintaining an extra trivial method. I'd lean towards the latter option, but either seems okay.